File: reference/dkp.rb
require 'json'

def dkp_log
  File.foreach("data/dkp.log-big").map { |line|
    amount, person, thing = line.strip.split(",")
    [ amount.to_i, person, thing ]
  }
end

standings = dkp_log.group_by { |trans| trans[1] }.map { |person, history|
  [ person, history.reduce(0) { |sum, trans| sum + trans[0] } ]
}.sort { |a, b| b[1] <=> a[1] }

puts standings.to_json

File: reference/dkp.scala

val standings = scala.io.Source.fromFile(new java.io.File("data/dkp.log-big"))
 .getLines()
 .map(line => {
   val Array(amount, person, thing) = line.split(',')
   (amount.toInt, person, thing)
 })
 .toSeq
 .groupBy(trans => trans._2)
 .toSeq
 .map(person => {
   person._1 -> person._2.foldLeft(0)((sum, trans) => sum + trans._1)
 })
 .sortWith((a, b) => b._2 < a._2)

println(standings)
